### PR TITLE
Add a number of model training enhancements: optimizer, activation func, learning rate, and kp filter

### PR DESCRIPTION
**Pull request by robertklee**
This is to update our model to attempt to resolve two problems: extreme unbounded predictions, slow training times

The main additions are:
* Optional change to activation function to use sigmoid instead of linear for the prediction layer. This has the benefit of bounding our output to [0,1], but has the drawback of potentially vanishing gradients for large x values (which has a gradient of 0).
* Change in our optimizer between RMSProp and Adam. Right now we use RMSProp, but Adam may be faster. 
* Increase learning rate by a factor of 10. We can add a scheduler to reduce it later on, but we are training way too slow. The pytorch repo uses 1e-2 where we use 5e-3. It used to be 5e-4
* Filter out annotations <= 4 keypoints. These are mostly garbage data and could adversely affect our model. We can still generalize pretty well since there's still a lot of high quality annotations >=5.

## Quality of life and dev experience

* adds a pretty big refactoring of hourglass.py because the parameters were getting out of hand. Everything is moved to the constructor now.
* makes our subset deterministic by seeding it with a state.
* automatically find highest epoch number if it is not provided
* change resume logic to use regex instead of hardcoded pattern

This knocks off 3 of our backlog items

### CODE REVIEW

File: evaluation.py [hunk 1]
```diff
@@ -27,7 +27,7 @@ def __init__(self, model_sub_dir, epoch, model_base_dir=DEFAULT_MODEL_BASE_DIR,
         if not os.path.exists(self.output_sub_dir):
             os.makedirs(self.output_sub_dir)
 
-        self.model_json, self.weights = util.find_resume_json_weights_str(model_base_dir, model_sub_dir, epoch)
+        self.model_json, self.weights, _ = util.find_resume_json_weights_str(model_base_dir, model_sub_dir, epoch)
```
Comment (Airborne95) [inline evaluation.py:30]: What does this do?

Comment (robertklee) [reply evaluation.py:30]: `util.find_resume_json_weights_str` returns a tuple of length 3 now so the `_` is required to unpack it properly

File: colab_train.ipynb [hunk 2]
```diff
@@ -220,17 +220,23 @@
         "outputId": "1f977904-3e1f-4e7d-c0ac-3d8511514c14"
       },
       "source": [
-        "!python3 train.py --epochs 100 --batch 12 --hourglass 4 --subset 1.0 --loss \"weighted_mse\" --augment \"none\"\n",
+        "!python3 train.py --epochs 100 --batch 16 --hourglass 4 --subset 1.0 --loss \"weighted_mse\" --augment \"light\"\n",
         "\n",
         "# To resume, add the following to the above command, and point it to the subdirectory of the model checkpoints. Ensure everything matches\n",
         "# --resume True --resume-epoch 42 --resume-subdir \"2021-03-14-15h-92m_batchsize_12_hg_4\"\n",
         "\n",
+        "# The options are below. NOTE that ALL options must match across training sessions:\n",
```
Comment (nicolepeverley) [inline colab_train.ipynb:228]: This is super helpful 💯

File: hourglass.py [hunk 3]
```diff
@@ -46,26 +56,33 @@ def load_and_filter_annotations(self, path_to_train_anns,path_to_val_anns, subse
...
-            train_df = train_df.sample(frac=subset)
+            train_df = train_df.sample(frac=subset, random_state=1)
         print(f"Train/Val dfs contains {len(train_df)}/{len(val_df)} anns")
         return train_df.reset_index(), val_df.reset_index()
 
-    def _start_train(self, batch_size, model_base_dir, epochs, initial_epoch, model_subdir, current_time, subset, loss_str, image_aug_str, pickle_name):
-        loss = get_loss_from_string(loss_str)
-        self._compile_model(loss)
+    def _start_train(self, batch_size, model_base_dir, epochs, initial_epoch, model_subdir, current_time, subset):
+        print(f'Loss function selected:               {self.loss_str}')
```
Comment (nicolepeverley) [inline hourglass.py:69]: Good idea to print the parameters!

File: util.py [hunk 4]
```diff
@@ -4,32 +4,63 @@
...
+        print('OutputActivation was not found in possible options.')
+        print('Available options are:')
+        available_activs = [name for name, member in OutputActivation.__members__.items()]
+        print(available_activs)
+        exit(1)
+    return activ
 
 def find_resume_json_weights_args(args):
-    return find_resume_json_weights_str(args.model_save, args.resume_subdir, args.resume_epoch)
+    args.resume_json, args.resume_weights, args.resume_epoch = find_resume_json_weights_str(args.model_save, args.resume_subdir, args.resume_epoch)
```
Comment (Airborne95) [inline util.py:30]: Just double checking, does this no longer need to be returned?

**Review by nicolepeverley — APPROVED** [review]
Nice clean up LGTM 👍

**Review by Airborne95 — APPROVED** [review]

File: util.py [hunk 5]
```diff
@@ -26,25 +26,41 @@ def validate_activation(activation):
         exit(1)
     return activ
 
-def find_resume_json_weights(args):
-    enclosing_dir = os.path.join(args.model_save, args.resume_subdir)
+def find_resume_json_weights_args(args):
+    args.model_save, args.resume_subdir, args.resume_epoch = find_resume_json_weights_str(args.model_save, args.resume_subdir, args.resume_epoch)
+
+def find_resume_json_weights_str(model_base_dir, model_subdir, resume_epoch):
```
Comment (Ckoelewyn) [inline util.py:32]: I like this

**Review by Ckoelewyn — APPROVED** [review]
Changes look good.

File: util.py (hunk 4 above)
Comment (robertklee) [reply util.py:30]: Yeah it was a bug. Shouldn't have been returned in the first place
